README.md
numform   
============


[![Build
Status](https://travis-ci.org/trinker/numform.svg?branch=master)](https://travis-ci.org/trinker/numform)
[![Coverage
Status](https://coveralls.io/repos/trinker/numform/badge.svg?branch=master)](https://coveralls.io/github/trinker/numform)
[![](https://cranlogs.r-pkg.org/badges/numform)](https://cran.r-project.org/package=numform)

**numform** contains tools to assist in the formatting of numbers and
plots for publication. Tools include the removal of leading zeros,
standardization of number of digits, addition of affixes, and a p-value
formatter. These tools combine the functionality of several 'base'
functions such as `paste()`, `format()`, and `sprintf()` into specific
use case functions that are named in a way that is consistent with
usage, making their names easy to remember and easy to deploy.

Installation
============

To download the development version of **numform**:

Download the [zip
ball](https://github.com/trinker/numform/zipball/master) or [tar
ball](https://github.com/trinker/numform/tarball/master), decompress and
run `R CMD INSTALL` on it, or use the **pacman** package to install the
development version:

    if (!require("pacman")) install.packages("pacman")
    pacman::p_load_current_gh("trinker/numform")
    pacman::p_load(tidyverse, gridExtra)


Table of Contents
============

-   [Installation](#installation)
-   [Contact](#contact)
-   [Available Functions](#available-functions)
-   [Demonstration](#demonstration)
    -   [Load Packages](#load-packages)
    -   [Numbers](#numbers)
    -   [Abbreviated Numbers](#abbreviated-numbers)
    -   [Commas](#commas)
    -   [Percents](#percents)
    -   [Dollars](#dollars)
    -   [Tables](#tables)
    -   [Plotting](#plotting)
    -   [Modeling](#modeling)

Contact
============


You are welcome to:    
- submit suggestions and bug-reports at: <https://github.com/trinker/numform/issues>    
- send a pull request on: <https://github.com/trinker/numform/>    
- compose a friendly e-mail to: <tyler.rinker@gmail.com>    


Available Functions
===================

Below is a table of available **numform** functions. Note that `f_` is
read as "format" whereas `fv_` is read as "format vector". The former
formats individual values in the vector while the latter uses the vector
to compute a calculation on each of the values and then formats them.
Additionally, all **numform** `f_` functions have a closure, function
retuning, version that is prefixed with an additional `f` (read "format
function"). For example, `f_num` has `ff_num` which has the same
arguments but returns a function instead. This is useful for passing in
to **ggplot2** `scale_x/y_type` functions (see [Plotting](#plotting) for
usage).

<!-- html table generated in R 3.5.1 by xtable 1.8-3 package -->
<!-- Thu Jan 17 08:14:34 2019 -->
<table>
<tr>
<td>
alignment
</td>
<td>
f_byte
</td>
<td>
f_latitude
</td>
<td>
f_peta
</td>
<td>
f_weekday_abbreviation
</td>
</tr>
<tr>
<td>
as_factor
</td>
<td>
f_celcius
</td>
<td>
f_list
</td>
<td>
f_pp
</td>
<td>
f_weekday_name
</td>
</tr>
<tr>
<td>
collapse
</td>
<td>
f_comma
</td>
<td>
f_list_amp
</td>
<td>
f_prefix
</td>
<td>
f_wrap
</td>
</tr>
<tr>
<td>
constant_months
</td>
<td>
f_data
</td>
<td>
f_logical
</td>
<td>
f_prop2percent
</td>
<td>
f_year
</td>
</tr>
<tr>
<td>
constant_months_abbreviation
</td>
<td>
f_data_abbreviation
</td>
<td>
f_longitude
</td>
<td>
f_pval
</td>
<td>
f_yotta
</td>
</tr>
<tr>
<td>
constant_quarters
</td>
<td>
f_date
</td>
<td>
f_mean_sd
</td>
<td>
f_quarter
</td>
<td>
f_zetta
</td>
</tr>
<tr>
<td>
constant_weekdays
</td>
<td>
f_degree
</td>
<td>
f_mega
</td>
<td>
f_replace
</td>
<td>
fv_num_percent
</td>
</tr>
<tr>
<td>
constant_weekdays_abbreviation
</td>
<td>
f_denom
</td>
<td>
f_mills
</td>
<td>
f_response
</td>
<td>
fv_percent
</td>
</tr>
<tr>
<td>
f_12_hour
</td>
<td>
f_dollar
</td>
<td>
f_month
</td>
<td>
f_sign
</td>
<td>
fv_percent_diff
</td>
</tr>
<tr>
<td>
f_abbreviation
</td>
<td>
f_exa
</td>
<td>
f_month_abbreviation
</td>
<td>
f_state
</td>
<td>
fv_percent_diff_fixed_relative
</td>
</tr>
<tr>
<td>
f_affirm
</td>
<td>
f_fahrenheit
</td>
<td>
f_month_name
</td>
<td>
f_suffix
</td>
<td>
fv_percent_lead
</td>
</tr>
<tr>
<td>
f_affix
</td>
<td>
f_giga
</td>
<td>
f_num
</td>
<td>
f_tera
</td>
<td>
fv_percent_lead_fixed_relative
</td>
</tr>
<tr>
<td>
f_bills
</td>
<td>
f_interval
</td>
<td>
f_num_percent
</td>
<td>
f_text_bar
</td>
<td>
fv_runs
</td>
</tr>
<tr>
<td>
f_bin
</td>
<td>
f_interval_right
</td>
<td>
f_ordinal
</td>
<td>
f_thous
</td>
<td>
glue
</td>
</tr>
<tr>
<td>
f_bin_right
</td>
<td>
f_interval_text
</td>
<td>
f_pad_zero
</td>
<td>
f_title
</td>
<td>
highlight_cells
</td>
</tr>
<tr>
<td>
f_bin_text
</td>
<td>
f_interval_text_right
</td>
<td>
f_parenthesis
</td>
<td>
f_trills
</td>
<td>
</td>
</tr>
<tr>
<td>
f_bin_text_right
</td>
<td>
f_kilo
</td>
<td>
f_percent
</td>
<td>
f_weekday
</td>
<td>
</td>
</tr>
</table>
<p class="caption">
<b><em>Available Formatting Functions</em></b>
</p>


Demonstration
=============

Load Packages
-------------

    if (!require("pacman")) install.packages("pacman")
    pacman::p_load_gh("trinker/numform")
    pacman::p_load(dplyr)

Numbers
-------

    f_num(c(0.0, 0, .2, -00.02, 1.122222, pi, "A"))

    ## [1] ".0"  ".0"  ".2"  "-.0" "1.1" "3.1" NA

Abbreviated Numbers
-------------------

    f_thous(1234)

    ## [1] "1K"

    f_thous(12345)

    ## [1] "12K"

    f_thous(123456)

    ## [1] "123K"

    f_mills(1234567)

    ## [1] "1M"

    f_mills(12345678)

    ## [1] "12M"

    f_mills(123456789)

    ## [1] "123M"

    f_bills(1234567891)

    ## [1] "1B"

    f_bills(12345678912)

    ## [1] "12B"

    f_bills(123456789123)

    ## [1] "123B"

...or auto-detect:

    f_denom(1234)

    ## [1] "1K"

    f_denom(12345)

    ## [1] "12K"

    f_denom(123456)

    ## [1] "123K"

    f_denom(1234567)

    ## [1] "1M"

    f_denom(12345678)

    ## [1] "12M"

    f_denom(123456789)

    ## [1] "123M"

    f_denom(1234567891)

    ## [1] "1B"

    f_denom(12345678912)

    ## [1] "12B"

    f_denom(123456789123)

    ## [1] "123B"

Commas
------

    f_comma(c(1234.12345, 1234567890, .000034034, 123000000000, -1234567))

    ## [1] "1,234.123"       "1,234,567,890"   ".000034034"      "123,000,000,000"
    ## [5] "-1,234,567"

Percents
--------

    f_percent(c(30, 33.45, .1), digits = 1)

    ## [1] "30.0%" "33.5%" ".1%"

    f_percent(c(0.0, 0, .2, -00.02, 1.122222, pi))

    ## [1] ".0%"  ".0%"  ".2%"  "-.0%" "1.1%" "3.1%"

    f_prop2percent(c(.30, 1, 1.01, .33, .222, .01))

    ## [1] "30.0%"  "100.0%" "101.0%" "33.0%"  "22.2%"  "1.0%"

    f_prop2percent(c(.30, 1, 1.01, .33, .222, .01), digits = 0)

    ## [1] "30%"  "100%" "101%" "33%"  "22%"  "1%"

    f_pp(c(.30, 1, 1.01, .33, .222, .01)) # same as f_prop2percent(digits = 0)

    ## [1] "30%"  "100%" "101%" "33%"  "22%"  "1%"

Dollars
-------

    f_dollar(c(0, 30, 33.45, .1))

    ## [1] "$0.00"  "$30.00" "$33.45" "$0.10"

    f_dollar(c(0.0, 0, .2, -00.02, 1122222, pi)) %>% 
        f_comma()

    ## [1] "$0.00"         "$0.00"         "$0.20"         "$-.02"        
    ## [5] "$1,122,222.00" "$3.14"

Sometimes one wants to lop off digits of money in order to see the
important digits, the real story. The `f_denom` family of functions can
do job.

    f_denom(c(12345267, 98765433, 658493021), prefix = '$')

    ## [1] "$ 12M" "$ 99M" "$658M"

    f_denom(c(12345267, 98765433, 658493021), relative = 1, prefix = '$')

    ## [1] "$ 12.3M" "$ 98.8M" "$658.5M"

Tables
------

Notice the use of the `alignment` function to detect the column
alignment.

    pacman::p_load(dplyr, pander)

    set.seed(10)
    dat <- data_frame(
        Team = rep(c("West Coast", "East Coast"), each = 4),
        Year = rep(2012:2015, 2),
        YearStart = round(rnorm(8, 2e6, 1e6) + sample(1:10/100, 8, TRUE), 2),
        Won = round(rnorm(8, 4e5, 2e5) + sample(1:10/100, 8, TRUE), 2),
        Lost = round(rnorm(8, 4.4e5, 2e5) + sample(1:10/100, 8, TRUE), 2),
        WinLossRate = Won/Lost,
        PropWon = Won/YearStart,
        PropLost = Lost/YearStart
    )


    dat %>%
        group_by(Team) %>%
        mutate(
            `%&Delta;WinLoss` = fv_percent_diff(WinLossRate, 0),
            `&Delta;WinLoss` = f_sign(Won - Lost, '<b>+</b>', '<b>&ndash;</b>')
            
        ) %>%
        ungroup() %>%
        mutate_at(vars(Won:Lost), .funs = ff_denom(relative = -1, prefix = '$')) %>%
        mutate_at(vars(PropWon, PropLost), .funs = ff_prop2percent(digits = 0)) %>%
        mutate(
            YearStart = f_denom(YearStart, 1, prefix = '$'),
            Team = fv_runs(Team),
            WinLossRate = f_num(WinLossRate, 1)
        ) %>%
        data.frame(stringsAsFactors = FALSE, check.names = FALSE) %>%
        pander::pander(split.tables = Inf, justify = alignment(.), style = 'simple')

<table>
<thead>
<tr class="header">
<th align="left">Team</th>
<th align="right">Year</th>
<th align="right">YearStart</th>
<th align="right">Won</th>
<th align="right">Lost</th>
<th align="right">WinLossRate</th>
<th align="right">PropWon</th>
<th align="right">PropLost</th>
<th align="right">%ΔWinLoss</th>
<th align="right">ΔWinLoss</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">West Coast</td>
<td align="right">2012</td>
<td align="right">$2.0M</td>
<td align="right">$350K</td>
<td align="right">$190K</td>
<td align="right">1.9</td>
<td align="right">17%</td>
<td align="right">9%</td>
<td align="right">0%</td>
<td align="right"><b>+</b></td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="right">2013</td>
<td align="right">$1.8M</td>
<td align="right">$600K</td>
<td align="right">$370K</td>
<td align="right">1.6</td>
<td align="right">33%</td>
<td align="right">20%</td>
<td align="right">-13%</td>
<td align="right"><b>+</b></td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="right">2014</td>
<td align="right">$ .6M</td>
<td align="right">$550K</td>
<td align="right">$300K</td>
<td align="right">1.8</td>
<td align="right">87%</td>
<td align="right">48%</td>
<td align="right">11%</td>
<td align="right"><b>+</b></td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="right">2015</td>
<td align="right">$1.4M</td>
<td align="right">$420K</td>
<td align="right">$270K</td>
<td align="right">1.6</td>
<td align="right">30%</td>
<td align="right">19%</td>
<td align="right">-13%</td>
<td align="right"><b>+</b></td>
</tr>
<tr class="odd">
<td align="left">East Coast</td>
<td align="right">2012</td>
<td align="right">$2.3M</td>
<td align="right">$210K</td>
<td align="right">$420K</td>
<td align="right">.5</td>
<td align="right">9%</td>
<td align="right">18%</td>
<td align="right">0%</td>
<td align="right"><b>–</b></td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="right">2013</td>
<td align="right">$2.4M</td>
<td align="right">$360K</td>
<td align="right">$390K</td>
<td align="right">.9</td>
<td align="right">15%</td>
<td align="right">16%</td>
<td align="right">86%</td>
<td align="right"><b>–</b></td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="right">2014</td>
<td align="right">$ .8M</td>
<td align="right">$590K</td>
<td align="right">$ 70K</td>
<td align="right">8.4</td>
<td align="right">74%</td>
<td align="right">9%</td>
<td align="right">811%</td>
<td align="right"><b>+</b></td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="right">2015</td>
<td align="right">$1.6M</td>
<td align="right">$500K</td>
<td align="right">$420K</td>
<td align="right">1.2</td>
<td align="right">30%</td>
<td align="right">26%</td>
<td align="right">-86%</td>
<td align="right"><b>+</b></td>
</tr>
</tbody>
</table>

    pacman::p_load(dplyr, pander)

    data_frame(
        Event = c('freezing water', 'room temp', 'body temp', 'steak\'s done', 'hamburger\'s done', 'boiling water', 'sun surface', 'lighting'),
        F = c(32, 70, 98.6, 145, 160, 212, 9941, 50000)
    ) %>%
        mutate(
            Event = f_title(Event),
            C = (F - 32) * (5/9)
        ) %>%
        mutate(
            F = f_degree(F, measure = 'F', type = 'string'),
            C = f_degree(C, measure = 'C', type = 'string', zero = '0.0')
        )  %>%
        data.frame(stringsAsFactors = FALSE, check.names = FALSE) %>%
        pander::pander(split.tables = Inf, justify = alignment(.), style = 'simple')

<table>
<thead>
<tr class="header">
<th align="left">Event</th>
<th align="right">F</th>
<th align="right">C</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">Freezing Water</td>
<td align="right">32.0°F</td>
<td align="right">0.0°C</td>
</tr>
<tr class="even">
<td align="left">Room Temp</td>
<td align="right">70.0°F</td>
<td align="right">21.1°C</td>
</tr>
<tr class="odd">

NEWS
NEWS
====

Versioning
----------

Releases will be numbered with the following semantic versioning format:

<major>.<minor>.<patch>

And constructed with the following guidelines:

* Breaking backward compatibility bumps the major (and resets the minor
  and patch)
* New additions without breaking backward compatibility bumps the minor
  (and resets the patch)
* Bug fixes and misc changes bumps the patch


numform 0.7.1 -
----------------------------------------------------------------

BUG FIXES

NEW FEATURES

MINOR FEATURES

CHANGES

IMPROVEMENTS



numform 0.6.4 - 0.7.0
----------------------------------------------------------------

BUG FIXES

* `f_ordinal` incorrectly converted the numbers 11-13 to '11st', '12nd' & '13rd'.
  Spotted by Jonathan Zadra (see #27).

* `f_dollar` had a `leading_zero` argument that was always TRUE.  This can be set
  to `FALSE` now.


MINOR FEATURES

* `f_num` picks up the argument `retain.leading.zero` to enable users to keep a
  leading zero before the decimal point if the value contains no whole number.
  Suggested by Dan Li.

* `f_percent` picks up an explicit `s` argument.  `f_dollar` picks up and
  explicit `p` argument.  This allows the percent and dollar sign to be set
  explicitly.  This can be useful if you need to escape such characters, for
  example, for use in LaTeX.  Suggested by Chris Snyder (see issue #26).


CHANGES

* `glue::collapse` was exported but has bee changed to `glue::glue_collapse`.



numform 0.6.1 - 0.6.3
----------------------------------------------------------------

NEW FEATURES

* `f_pad_left` and `f_pad_right` added to make padding a vector easy by computing
  the width from the data.

* `round2` added as an alternative to the default `round`'s design to round 1/2
  to the nearest even number.  THis version rounds 1/2 up.

* `round_any` added to allow rounding to non-whole number values.



numform 0.5.1 - 0.6.0
----------------------------------------------------------------

BUG FIXES

* `f_denom` functions did not properly handle scientific notation.  Spotted by
  Wilfried N'Guessan (see #23).

NEW FEATURES

* `f_text_bar` added to convert numeric vectors into an ASCII horizontal bar
  plot representation.

MINOR FEATURES

* `f_denom` picks up a functionality to the 100 trillions.

IMPROVEMENTS

* The sub-expression '(NaN|NA|Inf)' added to additional numeric
  (`additional.numeric` argument) of the `alignment()` function.



numform 0.4.1 - 0.5.0
----------------------------------------------------------------

BUG FIXES

* Functions that enable the argument `less.than.replace = TRUE` would result in
  the wrong solution if the input value was negative.  This has been corrected.

* `f_quarter` didn't respect the `space` or `max` arguments for some methods.

* `alignment` failed to align correctly if there were leading spaces.

NEW FEATURES

* `highlight_cells` added as a quick and dirty way to add cell font highlighting.

* The `collapse` and `glue` functions from the **glue** package are reexported
  for easy string manipulation.

* `f_bin` (a.k.a. `f_interval`) added for turning `"[3,5)"` form from `cut` into
  a more human readable for.  There are several different forms of `f_bin` so
  use `?f_bin` to see them all.

* `fv_percent_lead` added to work similar to `fv_percent_diff` but using the
  formula T_n/T_n-1.

* `fv_percent_lead_fixed_relative` & `fv_percent_diff_fixed_relative` added in
  order to compute `fv_percent_lead` & `fv_percent_diff` but keeping the
  comparison year as a constant rather than using the prior year in the vector
  of elements.

* `f_data` and friends (see `?f_data`) added to convert between data units.  For
  example, move from  10,000,000,000 (Bytes) to 10GB (Gigabytes) instead.

MINOR FEATURES

* `f_denom` family of functions picks up a `less.than.replace` argument to
  replace values that say zero with a less than sing followed by a one and then
  the denomination.  For example, in `c(2, 1234)` the digit 2 becomes "<1K".

* `f_denom` picks up a `mix.denom` argument to allow for the mixing of
  denominations.  This is useful for tables with a total row that is a
  denomination higher than the rest of the column.

* `f_pp` added to lower typing on `f_prop2percent(digits = 0)` (a commonly used
  function and setting) and to also change the default digits value to 0 as is
  typically used in plotting.

* `constant_weekdays_abbreviation` added to provide an abbreviated form of
  weekdays.

* `f_month_name` & `f_month_abbreviation` added to convert to full name and
  3 letter abbreviated month formats.

* `f_weekday_name` & `f_weekday_abbreviation` added to convert to full name and
  3 letter abbreviated weekday formats.


CHANGES

* The argument `less_than_replace` in `f_prop2percent` & `f_percent` has been
  replaced with `less.than.replace` to make dot case consistent for arguments
  across the package.




numform 0.3.1 - 0.4.0
----------------------------------------------------------------

BUG FIXES

* `f_12_hour` did not format correctly if the vector contained missing (`NA`)
  values; see issue #12.

NEW FEATURES

* `f_year`, `f_12_hour`, `f_month`, and `f_weekday` (methods based functions)
  did not previously have closure, function retuning, forms (i.e., `ff_` form)
  making them inconsistent in usage.  These functions all have a closure, function
  returning, form.

* `f_quarter` added for working with business quarters.

* `f_degree` and friends (`f_latitude`, `f_longitude`, `f_fahrenheit`, &
  `f_celcius`) added for working with degrees in scales and tables.

MINOR FEATURES

* `f_wrap` picks up an `equal.lines` argument to add additional lines to make
  each element have the same number of new line characters.  This is useful for
  legend spacing.

* `f_wrap` picks up a `collapse` argument to `paste` + `collapse` a vector before
  processing.  This is useful wrapping subtitles or other longer text blocks.


R/f_data.R
#' Convert and Abbreviate Units of Data.
#'
#' Convert numeric data to shorter form with unit abbreviations attached.  For
#' example, move from  10,000,000,000 (Bytes) to 10GB (Gigabytes) instead.
#'
#' @param x A vector of data units.
#' @param binary logical.  If \code{TRUE} the result uses binary conversion,
#' otherwise decimal conversion is used.
#' See \url{https://en.wikipedia.org/wiki/Binary_prefix} for additional
#' information on standards.
#' @param digits The number of digits to round to. .
#' @param pad.char A character to use for leading padding if lengths of output
#' are unequal.  Use \code{NA} to forgo padding.
#' @param less.than.replace logical.  If \code{TRUE} values lower than lowest
#' place value will be replaced with a less than sign followed by the
#' \code{integer} representation of the place value.  For example, if \code{"0GB"}
#' then replacement will be \code{"<1GB"}.
#' @param sep The separator to use between the number and data unit abbreviation.
#' @param mix.units logical.  If \code{TRUE} then units can be mixed.
#' Typically, this is not a good idea for the sake of comparison.  It is most
#' useful when there is a total row which is a sum of the column and this value's
#' unit exceeds the unit of the rest of the column.
#' @param from The starting unit.  Typically, this is assumed to be 'Bytes' ('B').
#' Must be one of c("Bit", "Byte", "Kilobyte", "Megabyte", "Gigabyte", "Terabyte",
#' "Petabyte", "Exabyte", "Zettabyte", "Yottabyte") or c("b", "B", "KB", "MB",
#' "GB", "TB", "PB", "EB", "ZB", "YB").  These are case sensitive.
#' @param \ldots ignored.
#' @param to The units to convert to.  See the \code{from} parameter for accepted
#' units.
#' @param suffix A suffix to use for the units at the end of the numeric string.
#' Typically the user will not interact with this argument.  Meant for internal
#' modularity of functions.
#' @return Returns a converted and abbreviated vector of units of data.
#' @export
#' @rdname f_data
#' @examples
#' \dontrun{
#' x <- c(NA, '3', '-', -233456789, -2334567890, 10^(0:10))
#' f_data(x)
#' f_data(x, pad.char = NA)
#' f_data(x, mix.units = TRUE)
#' f_data(x, mix.units = TRUE, binary = TRUE)
#' f_data(x, mix.units = TRUE, binary = TRUE, digits = 2)
#' f_byte(100000000, from = 'GB', binary = TRUE)
#' f_giga(10000000000)
#' f_giga(10000000000, suffix = 'Gb')
#'
#' library(tidyverse)
#' set.seed(15)
#' dat <- data_frame(
#'     bytes = round(rnorm(7, 1e7, 7.95e6), 0),
#'     days = constant_weekdays %>%
#'         as_factor()
#' )
#'
#' dat %>%
#'     mutate(
#'         data = f_data(bytes, less.than.replace = TRUE),
#'         weekday = f_weekday(days, distinct = TRUE) %>%
#'             as_factor()
#'     )
#'
#' dat %>%
#'     mutate(days = days %>% as_factor()) %>%
#'     ggplot(aes(days, bytes, group = 1)) +
#'         geom_line() +
#'         geom_point() +
#'         scale_y_continuous(labels = f_data) +
#'         scale_x_discrete(labels = ff_weekday(distinct = TRUE))
#' }
f_data <- function(x, binary = FALSE, digits = 0, pad.char = ' ',
    less.than.replace = FALSE, sep = '', mix.units = FALSE, from = 'B', ...) {

    ## convert to bits
    x <- f_byte_conversion(as.numeric(x), from = from, to = 'b')
    from <- 'b'

    if (mix.units) {

        chars <- nchar(drop_sci_not(abs(x)))
        is_na <- is.na(x)
        chars[is_na] <- 1

        to <- data_conversion[['Abbreviation']][unlist(lapply(chars, function(x) max(which(x > data_conversion[['Chars']]))))]

        out <- unlist(Map(function(y, u){

            if (is.na(y)) return(NA)

            f_data_default(
                x = y, to = u, binary = binary, digits = digits,
                suffix =  f_data_abbreviation(u),
                from = from, sep = sep
            )

        }, x, to))

        if (!is.na(pad.char)) {

            chars <- nchar(out)
            chars <- max(chars, na.rm = TRUE) - chars
            out <- paste0(unlist(lapply(chars, function(x) {
                    if (is.na(x)) return('')
                    paste(rep(pad.char, x), collapse = '')
                })), out)

        }

        out[is_na] <- NA

        out

    } else {
        chars <- nchar(drop_sci_not(max(abs(rm_na(x)))))
        to <- data_conversion[['Abbreviation']][max(which(chars > data_conversion[['Chars']]))]

        f_data_default(x = x, to = to, binary = binary, digits = digits, suffix =  f_data_abbreviation(to),
            pad.char = pad.char, less.than.replace = less.than.replace, from = from, sep = sep
        )

    }

}


data_conversion <- read.table(text =
"Unit    Abbreviation    Binary     Decimal   Step  Chars
Bit      b     		      1           1      0      0
Byte     B	   			8           8      1      1
Kilobyte KB 	         1024        1000      2      4
Megabyte MB			   1024        1000      3      7
Gigabyte GB			   1024        1000      4     10
Terabyte TB			   1024        1000      5     13
Petabyte PB			   1024        1000      6     16
Exabyte  EB			   1024        1000      7     19
Zettabyte ZB		   1024        1000      8     22
Yottabyte  YB		   1024        1000      9     25",
    header = TRUE,
    stringsAsFactors = FALSE
)


f_byte_conversion <- function(x, from, to, binary = FALSE, ..){

    tocol <- ifelse(nchar(to) > 2, 'Unit', "Abbreviation")
    if (!to %in% data_conversion[[tocol]]) error_data_conv('to')

    fromcol <- ifelse(nchar(from) > 2, 'Unit', "Abbreviation")
    if (!from %in% data_conversion[[fromcol]]) error_data_conv('from')

    convcol <- ifelse(binary, 'Binary', 'Decimal')

    fromloc <- match(from, data_conversion[[fromcol]])
    toloc <- match(to, data_conversion[[tocol]])

    if (fromloc == toloc) return(x)
    if (fromloc > toloc) dirfun <- `*` else dirfun <- `/`


    convfactor <- prod(data_conversion[[convcol]][(fromloc:toloc)[-which.min(fromloc:toloc)]])

    dirfun(x, convfactor)

}

error_data_conv <- function(var){

    stop(sprintf(
        '`%s` must match one of:\n\n    (%s)\n\n -OR-    \n\n    (%s) ',
        var,
        paste(shQuote(data_conversion[['Unit']]), collapse = ', '),
        paste(shQuote(data_conversion[['Abbreviation']]), collapse = ', ')
    ), call. = FALSE)

}


f_data_default <- function(x, to, binary = FALSE, digits = 0,
    suffix = f_data_abbreviation(to),
    pad.char = ' ', less.than.replace = FALSE, from = 'B', sep = '', ...) {

    x <- as.numeric(x)
    nas <- is.na(x)

    out <- f_byte_conversion(x, to = to, from = from, binary = binary)
    out <- f_num(out, digits = digits, s = paste0(sep, suffix))


    if (isTRUE(less.than.replace)){
        out <- gsub(
            '(^\\.?0+)([^0-9].*$)',
            paste0('<', gsub('.$', '1', f_num(0, digits = digits)), '\\2'),
            out
        )
    }

    if (!is.na(pad.char)) {

        chars <- nchar(out)
        chars <- max(chars, na.rm = TRUE) - chars
        out <- paste0(unlist(lapply(chars, function(x) {
                if (is.na(x)) return('')
                paste(rep(pad.char, x), collapse = '')
            })), out)

    }

    out[nas] <- NA

    out
}


#binary_factors <- decimal_factors <- rep(NA, 10)
#for (i in seq_len(nrow(data_conversion))){
#
#    if (i == 1) {
#        binary_factors[i] <- decimal_factors[i] <- 1
#    } else {
#        binary_factors[i] <- binary_factors[i - 1] * data_conversion[['Binary']][i]
#        decimal_factors[i] <- decimal_factors[i - 1] * data_conversion[['Decimal']][i]
#    }
#
#}

drop_sci_not <- function(x, ...){

    x <- as.character(as.numeric(x))
    locs <- grepl('e\\+', x)
    x[locs] <- paste0(
        gsub('e\\+.+', '', x[locs]),
        unlist(lapply(gsub('^.+?e\\+', '', x[locs]), function(x){paste(rep('0', x), collapse = '')}))
    )
    x
}


#' @export
#' @include utils.R
#' @rdname f_data
ff_data <- functionize(f_data)



#' @description \code{f_byte} - Force the abbreviation to bytes
#' unit (B).
#' @export
#' @include utils.R
#' @rdname f_data
f_byte <- hijack(f_data_default, to = 'B')

#' @export
#' @include utils.R
#' @rdname f_data
ff_byte <- functionize(f_byte)




#' @description \code{f_kilo} - Force the abbreviation to kilobytes
#' unit (KB).
#' @export
#' @include utils.R
#' @rdname f_data
f_kilo <- hijack(f_data_default, to = 'KB')

#' @export
#' @include utils.R
#' @rdname f_data
ff_kilo <- functionize(f_kilo)




#' @description \code{f_mega} - Force the abbreviation to megabytes
#' unit (MB).
#' @export
#' @include utils.R
#' @rdname f_data
f_mega <- hijack(f_data_default, to = 'MB')

#' @export
#' @include utils.R
#' @rdname f_data
ff_mega <- functionize(f_mega)




#' @description \code{f_giga} - Force the abbreviation to gigabytes
#' unit (GB).
#' @export
#' @include utils.R
#' @rdname f_data
f_giga <- hijack(f_data_default, to = 'GB')

#' @export
#' @include utils.R
#' @rdname f_data
ff_giga <- functionize(f_giga)




#' @description \code{f_tera} - Force the abbreviation to terabytes
#' unit (TB).
#' @export
#' @include utils.R
#' @rdname f_data
f_tera <- hijack(f_data_default, to = 'TB')

#' @export
#' @include utils.R
#' @rdname f_data
ff_tera <- functionize(f_tera)




#' @description \code{f_peta} - Force the abbreviation to petabytes
#' unit (PB).
#' @export
#' @include utils.R
#' @rdname f_data
f_peta <- hijack(f_data_default, to = 'PB')

#' @export
#' @include utils.R
#' @rdname f_data
ff_peta <- functionize(f_peta)




#' @description \code{f_exa} - Force the abbreviation to exabytes
#' unit (EB).
#' @export
#' @include utils.R
#' @rdname f_data
f_exa <- hijack(f_data_default, to = 'EB')

#' @export
#' @include utils.R
#' @rdname f_data
ff_exa <- functionize(f_exa)




#' @description \code{f_zetta} - Force the abbreviation to zettabytes
#' unit (ZB).
#' @export
#' @include utils.R
#' @rdname f_data
f_zetta <- hijack(f_data_default, to = 'ZB')

#' @export
#' @include utils.R
#' @rdname f_data
ff_zetta <- functionize(f_zetta)




#' @description \code{f_yotta} - Force the abbreviation to yottabytes
#' unit (YB).
#' @export
#' @include utils.R
#' @rdname f_data
f_yotta <- hijack(f_data_default, to = 'YB')

#' @export
#' @include utils.R
#' @rdname f_data
ff_yotta <- functionize(f_yotta)


#' Convert Data (byte) Labels to an Abbreviated Form
#'
#' Convert a data label such as Gigabyte to an abbreviated form
#' like 'GB'.
#'
#' @param x A vector of data labels.  One of:
#" c("Bit", "Byte", "Kilobyte", "Megabyte", "Gigabyte", "Terabyte",
#' "Petabyte", "Exabyte", "Zettabyte", "Yottabyte") ignoring case or
#' retaining c("b", "B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB")
#' with proper case.
#' @param \ldots ignored.
#' @return Returns avector of abbreviated data units.
#' @export
#' @examples
#' x <- c("Exabyte", "terabyte", "ZB", "PetaByte", "KB", "byte", "Gigabyte",
#' "Bit", "GB", "b")
#'
#' f_data_abbreviation(x)
f_data_abbreviation <- function(x, ...){

    out <- data_conversion[['Abbreviation']][match(tolower(x), tolower(data_conversion[['Unit']]))]
    out[is.na(out)] <- data_conversion[['Abbreviation']][match(x[is.na(out)], data_conversion[['Abbreviation']])]
    out
}





## to <- 'GB'
## x <- c(NA, '3', '-', -233456789, -2334567890, 10^(0:10))
##
##
## sprintf("%.100f", f_byte_conversion(1, from = 'GB', to = 'B'))
## sprintf("%.100f", f_byte_conversion(x, 'B', 'GB'))
## sprintf("%.100f", f_byte_conversion(x, 'B', 'GB', binary = TRUE))
## sprintf("%.100f", f_byte_conversion(x, from = 'B', to = 'b'))
##
## f_data_default(x, 'GB')
## f_data_default(x, 'GB', less.than.replace = TRUE)
## f_data_default(x, 'GB', sep = ' ')
## f_data_default(x, 'GB', less.than.replace = TRUE, sep = ' ')
## f_data_default(x, 'GB', less.than.replace = TRUE, sep = ' ', pad.char = ' ')
##
## hijack <- numform:::hijack
## functionize <- numform:::functionize
## hijack(f_data_default, to = 'GB')
## rm_na <- numform:::rm_na


##===============================================================================
## ## USED TO MAKE THE f_xxx & ff_xxx in a modular way
## cat(sprintf(
## paste0(
## "#' @description \\code{f_%s} - Force the abbreviation to %ss\n#' unit (%s).\n#' @export\n#' @include utils.R\n#' @rdname f_data\nf_%s <- hijack(f_data_default, to = '%s')\n\n",
##
## "#' @export\n#' @include utils.R\n#' @rdname f_data\nff_%s <- functionize(f_%s)\n\n\n\n"
##
## ),
##
## tolower(gsub('byte', '',  data_conversion[[1]])),
## tolower(data_conversion[[1]]),
## data_conversion[[2]],
## tolower(gsub('byte', '',  data_conversion[[1]])),
## data_conversion[[2]],
## tolower(gsub('byte', '',  data_conversion[[1]])),
## tolower(gsub('byte', '',  data_conversion[[1]]))
##
## ), sep='\n', file = 'clipboard')

